### DISCUSSION TITLE
考虑更新Wiki中的Docker方案？

### DISCUSSION THREAD

**Discussion by whentojump** [sjtug/SJTUThesis#872] [灌水区]
个人感觉MiKTeX镜像构建不太及时，且作者维护不积极（），容易遇到问题，见[Docker Hub](https://hub.docker.com/r/miktex/miktex)、[仓库](https://github.com/MiKTeX/docker-miktex)

或许可以把[Wiki中相关章节](https://github.com/sjtug/SJTUThesis/wiki/%E6%9C%AC%E5%9C%B0%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E#%E4%BD%BF%E7%94%A8%E9%A2%84%E7%BC%96%E8%AF%91%E7%9A%84-docker-%E9%95%9C%E5%83%8F%E7%BC%96%E8%AF%91%E6%A8%A1%E7%89%88)更新成[TeX Live](https://hub.docker.com/r/texlive/texlive)之类的镜像？

**Comment by LogCreative** [comment]
实际上，正如 Wiki 中指出的，MiKTeX 会自动更新需要的包，MiKTeX 的镜像仅仅是一个用来获取更新的壳子，所以可以理解为这个壳子不用频繁更新，包还是会拉最新的版本（

当然，非常欢迎提供 TeX Live 镜像的说明，因为 TeX Live 省事更稳定，所有的包安装一步到位，只是镜像体积相较于 MiKTeX 偏大（2GB vs 200MB）。

**Comment by LogCreative** [comment]
实际上，目前 Docker 在本项目 CI 中使用，也就是 

https://github.com/sjtug/SJTUThesis/blob/700093d999dc6fe212aa280d04aee6c120470f3b/.github/workflows/build.yml#L19

中使用的 [xu-cheng/texlive-action](https://github.com/xu-cheng/texlive-action)，这里面会拉取基于 Alpine Linux 的 TeX Live 镜像。可是，这个镜像经常会有一些问题，运行 XeLaTeX 时有一定的概率不稳定，造成了隔壁项目很多次的停摆，经常过几天自动恢复，见 https://github.com/sjtug/SJTUBeamer/pull/62#issuecomment-921808902 ，详细讨论见 https://github.com/sjtug/SJTUBeamer/pull/78 。所以有一个使用 TeX Live 原始镜像的配置也会对我们改进 CI 有一定的启示。如果能够做到最小化镜像，即只包含相关宏包，可能就比较理想了，详见 https://github.com/CTeX-org/ctex-kit/blob/master/.github/workflows/test.yml 。

**Comment by whentojump** [comment]
感谢答复！

> 所以可以理解为这个壳子不用频繁更新，包还是会拉最新的版本

这一点我觉得可能并不是这样，等我有空了提供一些出问题的例子（不过这个在这里就扯远了hh，应该在MiKTeX那里讨论（（您也可以参考下那边被bot自动关掉的若干个issue

> 欢迎提供 TeX Live 镜像的说明

嗯嗯，回头我打个草稿请各位指正一下，我相关知识也不深入哈，毕竟是较大的改动了怕误人子弟（其实我现在也就是起一个daemon当小虚拟机用了

> 改进 CI

或许[这个](https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/gitlab/ci/templates/LaTeX.gitlab-ci.yml)里面的几个第三方镜像能提供一些启发？不过咱们是中文，且宏包不一样可能也难以直接用上

您链接的几个issue/pr我也会学习一下！

**Reply by LogCreative** [reply]
> 那边被bot自动关掉的若干个issue

这个我还真没注意😂


> 不过咱们是中文，且宏包不一样可能也难以直接用上

这个没关系的，TeX Live 中的包是全的，包括中文相关宏包。主要是看镜像的构建时间是不是比较近，否则 TeX Live 一年更新一次，一旦过时可能会有无法更新宏包的麻烦。

**Reply by whentojump** [reply]
你好，我在MiKTeX中的具体问题发在这里了：https://github.com/MiKTeX/docker-miktex/issues/35 ，有时间可以帮忙看下是否可复现、甚至看看问题究竟出在哪:heart:

然后我写了个草稿：https://gist.github.com/whentojump/8618fe08adc572841c72aea0deb75f9d ，只是本人的用法，写着才意识到一些问题，还请各位提出建议！

1. 我只在Linux上测试了
2. 一些第三方镜像通常包括（1）一次性使用，和（2）后台进程两种脚本，所以我这边也包括了这两种方法；如果要和MiKTeX一节保持一致，可以删掉第二种
3. 一个比较致命的问题，TeX Live官方镜像（应该？）只有`root`用户，可能会造成麻烦乃至损害（我自己也就凑合用了，一直都没注意到这个事:joy:）。也许还是需要一个带普通用户的第三方镜像？

**Reply by LogCreative** [reply]
MiKTeX 的镜像现在确实不能用了，我在那边也回复了。

1. TeX Live 的部分，我在 macOS M1 上跑需要在 `run` 的时候加参数 `--platform linux/amd64`，可以提示一下。
2. 属于 docker 的通用用法，后台进程这种方法可以写个提示，可能很多人主要是一次性使用。
3. `root` 用户编译 TeX 文档一般没什么问题，只要编译时不打开 `-shell-escape` 开关都不会导致 TeX 运行系统命令。

**Reply by whentojump** [reply]
非常感谢！这两天我会把Wiki更新一下。除了已经讨论过的，再放个这个串的链接好了:smile:

**Reply by whentojump** [reply]
@LogCreative 抱歉鸽了一阵（

已经更新进Wiki里面了。想了想，最后还是和前面的gist换了个说法，感觉这样改动会小一点、不容易出错。请帮忙检查/指正！

**Reply by LogCreative** [reply]
Good job! 实际上你完善后可以把 TeX Live 相关信息写上的 没有问题👍
